### ISSUE TITLE
Timestamp incompatibility between firebase-admin and firebase/testing

### ISSUE THREAD

**Issue by AlexArlievsky** [firebase/firebase-js-sdk#3105]
### [REQUIRED] Describe your environment

  * Operating System version: Windows 10
  * Browser version: Chrome 81.0.4044.138 64 bit
  * Firebase SDK version: 8.4.0
  * Firebase Product: firestore

### [REQUIRED] Describe the problem
When using firebase/testing to run local tests, I encounter the following problem:
when I attempt to store document containing timestamp field I get an error:
"Function DocumentReference.set() called with invalid data. Unsupported field value: a custom object (found in field encrypted_data.registrationDate)"
encrypted_data.registrationDate is the field of type Timestamp.

#### Steps to reproduce:
Create db connection to local simulators using '@firebase/testing'
Create timestamp value using  FirebaseFirestore.Timestamp  from 'firebase-admin';
Store document with timestamp value.

Same code works when using  db connection from 'firebase-admin', both with local simulators and with cloud production environment.

Non-timestamp fields are stored normally.

#### Relevant Code:

The db connection and timestamps are obtained as follows:
```javascript
import * as firebaseTest from '@firebase/testing'
import admin from 'firebase-admin';
export class ServiceDirectoryTest {
////other stuff
private async initializeFirebase() {
        this.firebaseApp = firebaseTest.initializeAdminApp({ projectId: this.projectId })
        this.prodAdminApp = admin.initializeApp()   
    }
db(): FirebaseFirestore.Firestore {
        let db :any = this.firebaseApp.firestore()
        //let db = this.prodAdminApp.firestore()
        return db;
    }

now(): FirebaseFirestore.Timestamp {
        return admin.firestore.Timestamp.now()        
    }
}
```

**Comment by dconeybe** [comment]
Hi @AlexArlievsky. The `initializeAdminApp()` function returns a [firebase.app.App](https://firebase.google.com/docs/reference/js/firebase.app.App) object from the web client SDK, the same object type returned from `initializeTestApp()`. The only difference is that the object returned from `initializeAdminApp()` is tweaked via some internal hackery to emulate the admin sdk (e.g. security rules do not apply): [packages/testing/src/api/index.ts](https://github.com/firebase/firebase-js-sdk/blob/230cd489a4f903fb36d475adf1da408ccd1c7bff/packages/testing/src/api/index.ts#L112). Therefore, you need to use the [Timestamp](https://firebase.google.com/docs/reference/js/firebase.firestore.Timestamp) class from the web client SDK and not the admin SDK. This also applies to other types with shared names but different implementations, like [FieldValue](https://firebase.google.com/docs/reference/js/firebase.firestore.FieldValue).

So your code above would need to change to something like this:

```ts
import * as firebaseTest from '@firebase/testing'
import * as firebase from 'firebase'
import '@firebase/firestore'

...

now(): firebase.firestore.Timestamp {
        return firebase.firestore.Timestamp.now()
    }
}
```

Note that, despite its name, the `@firebase/testing` package is specifically geared towards testing rules and not a general-purpose means for connecting to the emulators. See https://firebase.google.com/docs/emulator-suite/connect_and_prototype for details. To tell the admin sdk to connect to the emulator it's easiest to use the `FIRESTORE_EMULATOR_HOST` environment variable and set it to `localhost:8080`.

**Comment by AlexArlievsky** [comment]
Thanks.
I tried to use both:
```
 this.firebaseApp = firebaseTest.initializeAdminApp({ projectId: this.projectId })
        let sa = {
            projectId: this.projectId, clientEmail: serviceAccount.client_email,
            privateKey: serviceAccount.private_key
        }
        // databaseURL: "localhost:8080",
        this.prodAdminApp = admin.initializeApp({ databaseURL: "localhost:8080", credential: admin.credential.cert(sa)})
        let db = this.prodAdminApp.firestore()
        db.settings({
            host: "localhost:8080",
            ssl: false
        });

This code is part of unit test initialization, and is executed outside of the firebase (GOOGLE_APPLICATION_CREDENTIALS points to the json file with admin level credentials. Same file is loaded and used in this snipped as serviceAccount).

Writes done through prodAdminApp  (real admin SDK) arrive to the local emulator, but are rejected due to security rules (looks like they run in the context of non-authenticated user, and not admin).

Writes done through firebaseApp (testing admin SDK) work OK (ignore rules and are executed by local emulator).

If I remove redirection to local emulator (locahost:8080), and use prodAdminApp writes are executed against cloud environment and work in admin mode (i.e. ignoring rules).

What is the correct way to have admin level access when working with local emulator outside of firebase (i.e. not as part of locally running Cloud Functions) ?

**Comment by yuchenshi** [comment]
Hi @AlexArlievsky , I'm sorry that you have to go through this hunt. The short answer to your last question is to use real admin SDKs to connect to the emulators. We're working on adding this bit to the documentation.

If I understand you correctly, that approach however did not work for you:

> Writes done through prodAdminApp (real admin SDK) arrive to the local emulator, but are rejected due to security rules (looks like they run in the context of non-authenticated user, and not admin).

This is unexpected and may be an issue with the SDK. The Admin SDK should be able to send over your credentials to the Firestore Emulator, which should then assume Admin privileges and skip security rules evaluation.

But to dig a bit further, we'd need a minimal repro. Would you mind putting up a gist with a self-contained Node.js script that shows how you load the service account and certificates? There is no need to include real private keys -- we're more interested in the exact way the SDK is initialized.

As a temporary workaround, please try to run your script this way:

```bash
FIREBASE_DATABASE_EMULATOR_HOST=localhost:8080 node your-admin-script.js
```

In other words, set the `FIREBASE_DATABASE_EMULATOR_HOST` environment variable to `localhost:8080` when running your script. Let me know if that works for you.

**Comment by AlexArlievsky** [comment]
Thanks.
I've created gist as you requested. 
https://gist.github.com/AlexArlievsky/7b1519ddbfbcd0ee00430185d67c851b
I assume your stuff is working and it is some stupid mistake on my side, and will be grateful if you could help me to sort it out.
10x.

**Comment by yuchenshi** [comment]
@AlexArlievsky  I cannot reproduce this. WIth your script, I can successfully write to prod and emulator.

There must be something that is different from your setup to my setup, and let's see:

* What is your Node.js version? (v8.16.0 and v10.16.3 work for me.)
* Did you upgrade `firebase-admin`? The latest is `firebase-admin@8.12.1`. Try `npm i -S firebase-admin@8.12.1`.
* Are you using a credentials file with the right format? I got mine from https://firebase.corp.google.com/project/_/settings/serviceaccounts/adminsdk and you should also try with a fresh one from there.

**Comment by AlexArlievsky** [comment]
node: v12.16.2 
firebase-admin: 8.12.1 (from package.json, also run your command)
Credentials are from firebase console, work when connecting to the cloud, and are in format:
```
{
  "type": "service_account",
  "project_id": "projectName",
  "private_key_id": "GUID",
  "private_key": "-----BEGIN PRIVATE KEY---------END PRIVATE KEY-----\n",
  "client_email": "firebase-adminsdk-XXX@XXXX.iam.gserviceaccount.com",
  "client_id": "some_number",
  "auth_uri": "https://accounts.google.com/o/oauth2/auth",
  "token_uri": "https://oauth2.googleapis.com/token",
  "auth_provider_x509_cert_url": "https://www.googleapis.com/oauth2/v1/certs",
  "client_x509_cert_url": "https://www.googleapis.com/robot/v1/metadata/x509/firebase-adminsdk-XXX%40XXXX.iam.gserviceaccount.com"
}

I also just created and downloaded new key file (just to be sure) - same behaviour.

**Comment by yuchenshi** [comment]
> "TypeError: Channel credentials must be a ChannelCredentials object"

This looks a lot like https://github.com/firebase/firebase-tools/issues/2207 and tl;dr it's highly likely to be caused by grpc dependency issue.

Try removing your `node_modules` directory (and maybe also `package-lock.json`) and `npm i` again. If things still don't work, attach `npm ls` output first, and then try again with a fresh new directory `npm init && npm i -S firebase-admin@8.12.1`.

**Comment by AlexArlievsky** [comment]
deleting package-lock.json and running npm i solved this problem. Now I'll try to modify my unit tests to work through admin sdk, as it should be. 
Thanks for your assistance.

**Comment by AlexArlievsky** [comment]
Update:
After grpc problem was resolved, I observed the original behaviour in unit tests - access denied.
Having working standalone test allowed quick finding of the root cause:
My attempt to make admin SDK to work with emulators using example from web sdk
```
db.settings({
            host: "localhost:8080",
            ssl: false
        });
```

succeeded to route calls to the emulator, but somehow lost the admin credentials in the process.

Apparently the only way to make admin SDK to work with emulator is using environment variable:
FIRESTORE_EMULATOR_HOST

As a low-priority feature I'd like to have a programmatic way to select which environment is used for each app / SDK instance (part of my test is to pull & mix data from different environments, different projects etc), but environment variable is certainly much better than nothing.

**Comment by yuchenshi** [comment]
The admin credentials issue when using `db.settings` sounds like a real bug either in the Node.js Admin SDK or Firestore Emulator. We're trying to find the root cause now.

**Comment by yuchenshi** [comment]
@AlexArlievsky FYI we've fixed the usage without environment variable. Feel free to upgrade and give it another try!
